src/server/game/Entities/Player/SceneMgr.cpp

```diff
@@ -0,0 +1,208 @@
+    sScriptMgr->OnSceneTrigger(GetPlayer(), sceneInstanceID, sceneTemplate, triggerName);
+}
+
+void SceneMgr::OnSceneCancel(uint32 sceneInstanceID)
+{
+    if (!HasScene(sceneInstanceID))
+        return;
+
+    if (m_isDebugScenes)
+        ChatHandler(GetPlayer()->GetSession()).PSendSysMessage(LANG_COMMAND_SCENE_DEBUG_CANCEL, sceneInstanceID);
+
+    SceneTemplate const* sceneTemplate = GetSceneTemplateFromInstanceId(sceneInstanceID);
+
+    // Must be done before removing aura
+    RemoveSceneInstanceId(sceneInstanceID);
+
+    if (sceneTemplate->SceneId != 0)
+        RemoveAurasDueToSceneId(sceneTemplate->SceneId);
+
+    sScriptMgr->OnSceneCancel(GetPlayer(), sceneInstanceID, sceneTemplate);
+}
+
+void SceneMgr::OnSceneComplete(uint32 sceneInstanceID)
+{
+    if (!HasScene(sceneInstanceID))
+        return;
+
+    if (m_isDebugScenes)
+        ChatHandler(GetPlayer()->GetSession()).PSendSysMessage(LANG_COMMAND_SCENE_DEBUG_COMPLETE, sceneInstanceID);
+
+    SceneTemplate const* sceneTemplate = GetSceneTemplateFromInstanceId(sceneInstanceID);
+
+    // Must be done before removing aura
+    RemoveSceneInstanceId(sceneInstanceID);
+
+    if (sceneTemplate->SceneId != 0)
+        RemoveAurasDueToSceneId(sceneTemplate->SceneId);
+
+    sScriptMgr->OnSceneComplete(GetPlayer(), sceneInstanceID, sceneTemplate);
+}
+
+bool SceneMgr::HasScene(uint32 sceneInstanceID, uint32 sceneScriptPackageId /*= 0*/) const
+{
+    auto itr = m_scenesByInstance.find(sceneInstanceID);
+    
+    if (itr == m_scenesByInstance.end())
+        return false;
+
+    return !sceneScriptPackageId || sceneScriptPackageId == itr->second->ScenePackageId;
+}
+
+void SceneMgr::AddInstanceIdToSceneMap(uint32 sceneInstanceID, SceneTemplate const* sceneTemplate)
+{
+    m_scenesByInstance[sceneInstanceID] = sceneTemplate;
+}
+
+void SceneMgr::CancelSceneByPackageId(uint32 sceneScriptPackageId)
+{
+    std::vector<uint32> instancesIds;
+
+    for (auto itr : m_scenesByInstance)
+        if (itr.second->ScenePackageId == sceneScriptPackageId)
+            instancesIds.push_back(itr.first);
+
+    for (uint32 sceneInstanceID : instancesIds)
+        CancelScene(sceneInstanceID);
+}
+
+void SceneMgr::RemoveSceneInstanceId(uint32 sceneInstanceID)
+{
+    m_scenesByInstance.erase(sceneInstanceID);
+}
+
+void SceneMgr::RemoveAurasDueToSceneId(uint32 sceneId)
+{
+    Player::AuraEffectList const& scenePlayAuras = GetPlayer()->GetAuraEffectsByType(SPELL_AURA_SCENE_PLAY);
+    for (AuraEffect* scenePlayAura : scenePlayAuras)
+    {
+        if (static_cast<uint32>(scenePlayAura->GetMiscValue()) == sceneId)
+        {
+            GetPlayer()->RemoveAura(scenePlayAura->GetBase());
+            break;
+        }
+    }
+}
+
+SceneTemplate const* SceneMgr::GetSceneTemplateFromInstanceId(uint32 sceneInstanceID)
+{
+    auto itr = m_scenesByInstance.find(sceneInstanceID);
+
+    if (itr == m_scenesByInstance.end())
```

Comment:
there is a mix of codestyles, sometimes itr is compared with == .end(), sometimes it's compared with != .